hw/dv/dpi/common/tcp_server/tcp_buffer.c

```diff
@@ -0,0 +1,60 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "tcp_buffer.h"
+
+#include <stdlib.h>
+
+#define BUFSIZE 256
+
+struct tcp_buffer_struct {
+  unsigned int rptr;
+  unsigned int wptr;
+  char buf[BUFSIZE];
+};
+
+static bool tcp_buffer_full(struct tcp_buffer_struct *buf) {
+  if (buf->wptr >= buf->rptr) {
+    return (buf->wptr - buf->rptr) == (BUFSIZE - 1);
+  } else {
+    return (buf->rptr - buf->wptr) == 1;
+  }
+}
+
+static bool tcp_buffer_empty(struct tcp_buffer_struct *buf) {
+  return (buf->wptr == buf->rptr);
+}
+
+void tcp_buffer_put_byte(struct tcp_buffer_struct *buf, char dat) {
+  bool done = false;
+  while (!done) {
+    if (!tcp_buffer_full(buf)) {
+      buf->buf[buf->wptr++] = dat;
+      buf->wptr &= (BUFSIZE - 1);
+      done = true;
+    }
+  }
+}
+
+bool tcp_buffer_get_byte(struct tcp_buffer_struct *buf, char *dat) {
+  if (tcp_buffer_empty(buf)) {
+    return false;
+  }
+  *dat = buf->buf[buf->rptr++];
+  buf->rptr &= (BUFSIZE - 1);
+  return true;
+}
+
+struct tcp_buffer_struct *tcp_buffer_init() {
+  struct tcp_buffer_struct *buf_new;
+  buf_new =
+      (struct tcp_buffer_struct *)malloc(sizeof(struct tcp_buffer_struct));
+  buf_new->rptr = 0;
+  buf_new->wptr = 0;
+  return buf_new;
+}
+
+void tcp_buffer_delete(struct tcp_buffer_struct *buf) { free(buf); }
```

Comment:
- Please rename the function to tcp_buffer_free().
- To be slightly safer to use, zero the pointer in the function:
   ```c
   void tcp_buffer_free(struct tcp_buffer_struct **buf) { free(*buf); *buf = NULL; }
    ```
    This allows you to call `tcp_buffer_free()` safely multiple times and ensures that checks like `buf == NULL` are actually giving the intended result after the buffer has been free'd.